pgml-cms/blog/semantic-search-in-postgres-in-15-minutes.md

```diff
@@ -0,0 +1,467 @@
+
+Let's demonstrate this by inserting 100,000 additional embeddings:
+
+!!! generic
+
+!!! code_block
+
+```postgresql
+INSERT INTO text_and_embeddings (text, embedding) 
+SELECT
+  md5(random()::text),
+  pgml.embed(
+    'mixedbread-ai/mxbai-embed-large-v1',
+    md5(random()::text)
+  ) 
+FROM generate_series(1, 100000);
+```
+
+!!!
+
+!!!
+
+Now trying our search engine again:
+
+!!! generic
+
+!!! code_block time="105.917 ms"
+
+```postgresql
+WITH embedded_query AS (
+    SELECT
+        pgml.embed('mixedbread-ai/mxbai-embed-large-v1', 'What is the pgml.transform function?', '{"prompt": "Represent this sentence for searching relevant passages: "}')::vector embedding
+)
+SELECT
+    text,
+    (
+        SELECT
+            embedding
+        FROM embedded_query) <=> text_and_embeddings.embedding cosine_distance
+FROM
+  text_and_embeddings
+ORDER BY cosine_distance
+LIMIT 1;
+```
+
+!!!
+
+!!! results
+
+```
+                                          text                                          |   cosine_distance   
+----------------------------------------------------------------------------------------+---------------------
+ The pgml.transform function is a PostgreSQL function for calling LLMs in the database. | 0.13467974993681486
+```
+
+!!!
+
+!!!
+
+This somewhat less than ideal performance can be fixed by indexing the embedding column. There are two types of indexes available in _pgvector_: IVFFlat and HNSW.
+
+IVFFlat indexes clusters the table into sublists, and when searching, only searches over a fixed number of sublists. In our example, if we were to add an IVFFlat index with 10 lists:
+
+!!! generic
+
+!!! code_block time="4989.398 ms"
+
+```postgresql
+CREATE INDEX ON text_and_embeddings
+USING ivfflat (embedding vector_cosine_ops)
+WITH (lists = 10);
+```
+
+!!!
+
+!!!
+
+and search again, we would get much better performance:
+
+!!! generic
+
+!!! code_block time="29.191 ms"
+
+```postgresql
+WITH embedded_query AS (
+    SELECT
+        pgml.embed('mixedbread-ai/mxbai-embed-large-v1', 'What is the pgml.transform function?', '{"prompt": "Represent this sentence for searching relevant passages: "}')::vector embedding
+)
+SELECT
+    text,
+    (
+        SELECT
+            embedding
+        FROM embedded_query) <=> text_and_embeddings.embedding cosine_distance
+FROM
+  text_and_embeddings
+ORDER BY cosine_distance
+LIMIT 1;
+```
+
+!!!
+
+!!! results
+
+```
+                                          text                                          |   cosine_distance   
+----------------------------------------------------------------------------------------+---------------------
+ The pgml.transform function is a PostgreSQL function for calling LLMs in the database. | 0.13467974993681486
+```
+
+!!!
+
+!!!
+
+We can see it is a massive speedup because we are only searching over 1/10th of the original vectors! 
```

Comment:
```suggestion
We can see it is a massive speedup because we are only comparing our input to 1/10th of the original vectors, instead of all of them! 
```